### PR TITLE
TPP dialect changes for VNNI Ops

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**

### CODE REVIEW

File: include/TPP/Dialect/Tpp/TppOps.td [hunk 1]
```diff
@@ -23,9 +23,9 @@ class StaticMemRefRankOf<list<Type> allowedTypes, list<int> ranks> :
          "::mlir::MemRefType">;
 
 def TppMemRef : StaticMemRefRankOf<[AnyFloat], [1, 2]>;
-def TppPackedMemrefInput : StaticMemRefRankOf<[AnyFloat], [1, 2, 3]>;
+def TppPackedMemrefInput : StaticMemRefRankOf<[AnyFloat], [3]>;
```
Comment (chelini) [inline include/TPP/Dialect/Tpp/TppOps.td:26]: should we use VNNI operand instead of packed?

File: include/TPP/Dialect/Tpp/TppOps.td [hunk 2]
```diff
@@ -138,6 +138,53 @@ def Tpp_MatmulOp : Tpp_Op<"matmul"> {
...
+
+  let hasVerifier = 1;
+}
+
+//===----------------------------------------------------------------------===//
+// VNNI_MatmulOp
+//===----------------------------------------------------------------------===//
+
+def Tpp_VNNI_MatmulOp : Tpp_Op<"vnni_matmul"> {
+  let summary = "Performs matrix multiplication of two inputs, with the first operand in VNNI format.";
```
Comment (chelini) [inline include/TPP/Dialect/Tpp/TppOps.td:173]: If it is possible can we format better the `summary`?

File: include/TPP/Dialect/Tpp/TppOps.td [hunk 3]
```diff
@@ -138,6 +138,53 @@ def Tpp_MatmulOp : Tpp_Op<"matmul"> {
...
+
+  let description = [{
+    The `tpp.vnni_matmul` mirrors `vnni.matmul`.
```
Comment (chelini) [inline include/TPP/Dialect/Tpp/TppOps.td:176]: what is `vnni.matmul`? We have only `tpp.vnni_matmul`, right?

File: include/TPP/Dialect/Tpp/TppOps.td [hunk 4]
```diff
@@ -138,6 +138,53 @@ def Tpp_MatmulOp : Tpp_Op<"matmul"> {
...
+
+    Example:
+
+    ```mlir
+ 
+    tpp.vnni_matmul ins(%1: memref<2x2x2xf32>, %2: memref<2x2xf32>) 
+               out(%3: memref<4x2xf32>)
```
Comment (chelini) [inline include/TPP/Dialect/Tpp/TppOps.td:183]: format better pls.

File: include/TPP/Dialect/Tpp/TppOps.td [hunk 5]
```diff
@@ -214,4 +261,56 @@ def Tpp_BrgemmOp : Tpp_Op<"brgemm"> {
   let hasVerifier = 1;
 }
 
+
+//===----------------------------------------------------------------------===//
+// VNNI_BrgemmOp
+//===----------------------------------------------------------------------===//
+
+def Tpp_VNNI_BrgemmOp : Tpp_Op<"vnni_brgemm"> {
+  let summary = "Performs batch reduced matrix multiplication of two inputs with the first input in VNNI format";
```
Comment (chelini) [inline include/TPP/Dialect/Tpp/TppOps.td:270]: format better pls

File: lib/TPP/ConvertTppToXsmm.cpp [hunk 6]
```diff
@@ -440,7 +521,9 @@ void mlir::tpp::populateTppToXsmmPatterns(RewritePatternSet &patterns) {
                ConvertTppReluOp,
                ConvertTppAddOp,
                ConvertTppMatmulOp,
-               ConvertTppBrgemmOp>(patterns.getContext());
+	       ConvertTpp_VNNI_MatmulOp,
+               ConvertTppBrgemmOp,
+		ConvertTpp_VNNI_BrgemmOp>(patterns.getContext());
```
Comment (chelini) [inline lib/TPP/ConvertTppToXsmm.cpp:526]: format better pls

File: lib/TPP/Dialect/Tpp/TppOps.cpp [hunk 7]
```diff
@@ -60,28 +60,23 @@ LogicalResult IdentityOp::verify() {
...
   int64_t m = shapeC[0];
   int64_t n = shapeC[1];
   int64_t k = shapeA[1];
   // Verify C(m, n) = A(m, k) B(k, n)
-  if (shapeB[0] != k || shapeB[1] != n)
+  if (shapeB[0] != k || shapeB[1] != n || shapeA[0] != m)
     return false;
-  return ((isPackedBF16 && shapeA[0] * shapeA[2] == m) ||
-          (!isPackedBF16 && (shapeA[0] == m) && (shapeA[1] == k)));
+  return true;
```
Comment (chelini) [inline lib/TPP/Dialect/Tpp/TppOps.cpp:79]: yes, thanks a lot! Simplify this code is perfect.

File: lib/TPP/Dialect/Tpp/TppOps.cpp [hunk 8]
```diff
@@ -165,3 +149,76 @@ LogicalResult AddOp::verify() {
...
+}
+
+static bool verifyVNNIMatmulOperandsDims(ArrayRef<int64_t> shapeA,
+                                         ArrayRef<int64_t> shapeB,
+                                         ArrayRef<int64_t> shapeC) {
+  int64_t m = shapeC[0];
+  int64_t n = shapeC[1];
+  int64_t k = shapeA[1];
+  // Verify C(m, n) = A(m, k) B(k, n)
```
Comment (chelini) [inline lib/TPP/Dialect/Tpp/TppOps.cpp:171]: this comment may not be true anymore.

File: lib/TPP/Dialect/Tpp/TppOps.cpp [hunk 9]
```diff
@@ -165,3 +149,76 @@ LogicalResult AddOp::verify() {
...
+
+static bool verifyVNNIBRGemmShape(MemRefType memrefA, MemRefType memrefB,
+                                  MemRefType memrefC) {
+  if (memrefB.getRank() != 3 || memrefC.getRank() != 2 ||
+      memrefA.getRank() != 4)
+    return false;
+  return true;
+}
+
+// XXX: Changing the op semantics based on the type is so bad and brittle.
```
Comment (chelini) [inline lib/TPP/Dialect/Tpp/TppOps.cpp:207]: now we can drop this comment ;)

**Review by chelini — APPROVED** [review]
Perfect, thanks!

File: include/TPP/Dialect/Tpp/TppOps.td (hunk 3 above)
Comment (KavithaTipturMadhu) [reply include/TPP/Dialect/Tpp/TppOps.td:176]: I am adding vnni dialect, added the comment in anticipation of the op being added. I'll revert this for now.

File: include/TPP/Dialect/Tpp/TppOps.td (hunk 2 above)
Comment (chelini) [reply include/TPP/Dialect/Tpp/TppOps.td:173]: try with:
`let summary = [{
  bla bla
  bla bla
}];`

File: include/TPP/Dialect/Tpp/TppOps.td (hunk 5 above)
Comment (KavithaTipturMadhu) [reply include/TPP/Dialect/Tpp/TppOps.td]: This doesn't take a newline argument so I am forced to type the whole statement in one line.

**Comment by rengolin** [comment]
> @rengolin With this patch, ops (matmul and brgemm) for 3d input for matmul/4d for brgemm with bf16 type no longer works. Is it okay to assume that vnni_op is the default op type to be used for all tensors bf16 (with 3d input for matmul and 4d for brgemm)?

Correct. I think this gives a clear separation of concerns and we can focus on the transforms and shapes we need for each different type. If/when we get to the point where we want to merge/upstream, we can review this stance and refactor.

LGTM, thanks!

**Pull request merged**